### PR TITLE
Push0 rematerializer default sequence

### PR DESCRIPTION
**Pull request by nikola-matic**
This uses https://github.com/ethereum/solidity/pull/14469 as the base branch; alongside the rematerializer changes from the base PR, this one simply adds the rematerializer and unused pruner steps (`mu`) to the end of the default clean up sequence.

### CODE REVIEW

File: libsolidity/interface/OptimiserSettings.h [hunk 1]
```diff
@@ -59,7 +59,7 @@ struct OptimiserSettings
 		"]"
 		"jmul[jul] VcTOcul jmul";      // Make source short and pretty
 
-	static char constexpr DefaultYulOptimiserCleanupSteps[] = "fDnTOc";
+	static char constexpr DefaultYulOptimiserCleanupSteps[] = "fDnTOcm";
```
Comment (ekpyron) [inline libsolidity/interface/OptimiserSettings.h]: We should probably make it ``mu`` in the end to get rid of the dead variables, right?

Comment (nikola-matic) [reply libsolidity/interface/OptimiserSettings.h]: Sure, I'll add it in a separate commit so we can see the diff with `u` and without.

**Comment by nikola-matic** [comment]
<details><summary>Click for a table of gas differences</summary>

| File name                                                                 |   IR-optimized (%) |   Legacy-Optimized (%) |   Legacy (%) |
|---------------------------------------------------------------------------|--------------------|------------------------|--------------|
| byte_array_to_storage_cleanup.sol                                         |       -0.0983875   |            0.225641    |            0 |
| structs/struct_containing_bytes_copy_and_delete.sol                       |        0.00748604  |           -0.00224091  |            0 |
| structs/copy_substructures_from_mapping.sol                               |        0           |           -0.00246374  |            0 |
| structs/copy_to_mapping.sol                                               |       -0.00588307  |           -0.00245969  |            0 |
| structs/copy_substructures_to_mapping.sol                                 |       -0.000839839 |           -0.00242301  |            0 |
| structs/copy_from_mapping.sol                                             |       -0.0049388   |           -0.00246388  |            0 |
| viaYul/copy_struct_invalid_ir_bug.sol                                     |       -0.00973624  |            0           |            0 |
| various/create_calldata.sol                                               |        0.451271    |            0.145268    |            0 |
| various/selfdestruct.sol                                                  |       -0.00641814  |            0           |            0 |
| various/skip_dynamic_types_for_structs.sol                                |       -0.00672318  |            0           |            0 |
| various/senders_balance.sol                                               |       -0.512064    |            0           |            0 |
| various/value_insane.sol                                                  |       -0.372587    |            0           |            0 |
| various/address_code.sol                                                  |        0.242981    |            0.140759    |            0 |
| various/value_complex.sol                                                 |       -0.374447    |            0           |            0 |
| various/destructuring_assignment.sol                                      |       -0.00165404  |            0           |            0 |
| constructor/bytes_in_constructors_packer.sol                              |       -0.14167     |            0.0685495   |            0 |
| constructor/constructor_static_array_argument.sol                         |        0           |            0.00908085  |            0 |
| constructor/arrays_in_constructors.sol                                    |        0.00424017  |            0.0024742   |            0 |
| constructor/bytes_in_constructors_unpacker.sol                            |        0.402767    |            0.0835952   |            0 |
| externalContracts/prbmath_signed.sol                                      |       -0.0873192   |            0           |            0 |
| externalContracts/strings.sol                                             |        0.204943    |            0.0297846   |            0 |
| externalContracts/prbmath_unsigned.sol                                    |       -0.0501317   |            0           |            0 |
| externalContracts/base64.sol                                              |       -0.522076    |            0.0410007   |            0 |
| externalContracts/deposit_contract.sol                                    |        0.31383     |            0           |            0 |
| immutable/multi_creation.sol                                              |       -0.0024193   |            0           |            0 |
| functionCall/gas_and_value_basic.sol                                      |       -0.515062    |            0           |            0 |
| functionCall/creation_function_call_with_args.sol                         |       -0.446357    |            0           |            0 |
| functionCall/gas_and_value_brace_syntax.sol                               |       -0.515062    |            0           |            0 |
| functionCall/external_call_to_nonexisting.sol                             |       -1.03919     |            0           |            0 |
| functionCall/creation_function_call_with_salt.sol                         |       -0.445982    |            0           |            0 |
| functionCall/failed_create.sol                                            |       -3.29325     |            0           |            0 |
| abiEncoderV1/abi_encode_calldata_slice.sol                                |        0.0206811   |            0.0163291   |            0 |
| abiEncoderV1/struct/struct_storage_ptr.sol                                |        0.00480576  |            0           |            0 |
| salted_create/salted_create_with_value.sol                                |       -0.0118561   |            0           |            0 |
| abiEncoderV2/abi_encode_v2_in_function_inherited_in_v1_contract.sol       |        0.0180618   |            0.0126859   |            0 |
| abiEncoderV2/storage_array_encoding.sol                                   |       -0.0486704   |            0           |            0 |
| abiEncoderV2/calldata_array.sol                                           |        0.00157668  |            0           |            0 |
| abiEncoderV2/abi_encode_calldata_slice.sol                                |        0.0206811   |            0.0163291   |            0 |
| userDefinedValueType/calldata.sol                                         |        0.00408537  |            0           |            0 |
| libraries/internal_types_in_library.sol                                   |        0.0160549   |            0           |            0 |
| libraries/using_library_mappings_return.sol                               |       -0.00499168  |            0           |            0 |
| libraries/using_library_mappings_public.sol                               |       -0.00585079  |            0           |            0 |
| abiencodedecode/abi_decode_simple_storage.sol                             |       -0.000737126 |            0           |            0 |
| events/event_indexed_string.sol                                           |        0           |           -0.000267543 |            0 |
| events/event_emit_from_other_contract.sol                                 |       -0.00725523  |            0           |            0 |
| array/create_memory_array.sol                                             |        0.00238438  |            0           |            0 |
| array/array_storage_index_zeroed_test.sol                                 |        0.000293423 |            0           |            0 |
| array/arrays_complex_from_and_to_storage.sol                              |       -0.00160811  |            0           |            0 |
| array/function_array_cross_calls.sol                                      |        0.0823981   |            0           |            0 |
| array/invalid_encoding_for_storage_byte_array.sol                         |        0.00165147  |           -0.00164745  |            0 |
| array/bytes_length_member.sol                                             |        0.000905863 |           -0.00271299  |            0 |
| array/fixed_arrays_as_return_type.sol                                     |        0.175125    |            0           |            0 |
| array/array_memory_index_access.sol                                       |        0.00444247  |            0           |            0 |
| array/pop/byte_array_pop_masking_long.sol                                 |       -0.0129667   |            0           |            0 |
| array/pop/array_pop_uint16_transition.sol                                 |       -0.0101292   |            0           |            0 |
| array/pop/array_pop_uint24_transition.sol                                 |       -0.00317279  |            0           |            0 |
| array/copying/array_of_structs_containing_arrays_memory_to_storage.sol    |       -0.00219887  |            0           |            0 |
| array/copying/memory_dyn_2d_bytes_to_storage.sol                          |       -0.13253     |            0           |            0 |
| array/copying/array_copy_target_simple_2.sol                              |       -0.00685898  |            0           |            0 |
| array/copying/array_of_struct_memory_to_storage.sol                       |        0.00168976  |            0           |            0 |
| array/copying/array_copy_storage_storage_different_base_nested.sol        |       -0.0123831   |            0           |            0 |
| array/copying/bytes_inside_mappings.sol                                   |        0           |           -0.00207999  |            0 |
| array/copying/array_copy_target_leftover.sol                              |       -0.00765301  |            0           |            0 |
| array/copying/array_copy_storage_storage_different_base.sol               |       -0.000487698 |            0           |            0 |
| array/copying/copying_bytes_multiassign.sol                               |        0           |           -0.00450167  |            0 |
| array/copying/nested_array_of_structs_memory_to_storage.sol               |       -0.00194069  |            0           |            0 |
| array/copying/elements_of_nested_array_of_structs_memory_to_storage.sol   |       -0.00178983  |            0           |            0 |
| array/copying/array_nested_memory_to_storage.sol                          |        0.0014162   |            0           |            0 |
| array/copying/storage_memory_nested_from_pointer.sol                      |       -0.000971435 |            0           |            0 |
| array/copying/bytes_storage_to_storage.sol                                |       -0.229274    |           -0.000863745 |            0 |
| array/copying/copy_removes_bytes_data.sol                                 |        0.000563777 |           -0.00168964  |            0 |
| array/copying/calldata_array_dynamic_to_storage.sol                       |        0.00540726  |            0           |            0 |
| array/copying/copy_function_internal_storage_array.sol                    |        0.00571643  |            0           |            0 |
| array/copying/nested_array_element_storage_to_storage.sol                 |        0.00531703  |            0           |            0 |
| array/copying/calldata_array_to_mapping.sol                               |       -0.00143179  |            0           |            0 |
| array/copying/array_of_struct_calldata_to_storage.sol                     |        0.0108613   |            0           |            0 |
| array/copying/array_copy_storage_storage_static_dynamic.sol               |       -0.00081211  |            0           |            0 |
| array/copying/elements_of_nested_array_of_structs_calldata_to_storage.sol |        0.00821631  |            0           |            0 |
| array/copying/array_nested_calldata_to_storage.sol                        |        0.00371391  |            0           |            0 |
| array/copying/array_copy_nested_array.sol                                 |       -0.000144998 |            0           |            0 |
| array/copying/array_copy_different_packing.sol                            |        0.010094    |            0           |            0 |
| array/copying/array_copy_target_simple.sol                                |       -0.000365545 |            0           |            0 |
| array/copying/array_copy_cleanup_uint40.sol                               |        0.0025584   |            0           |            0 |
| array/copying/nested_array_of_structs_calldata_to_storage.sol             |        0.00886403  |            0           |            0 |
| array/copying/array_copy_calldata_storage.sol                             |       -0.00061698  |            0           |            0 |
| array/copying/storage_memory_packed_dyn.sol                               |        0.00364637  |            0           |            0 |
| array/copying/storage_memory_nested.sol                                   |       -0.000971435 |            0           |            0 |
| array/copying/nested_dynamic_array_element_calldata_to_storage.sol        |       -0.00121767  |            0           |            0 |
| array/copying/function_type_array_to_storage.sol                          |       -0.00318367  |            0           |            0 |
| array/copying/storage_memory_nested_bytes.sol                             |       -0.00493247  |            0           |            0 |
| array/delete/bytes_delete_element.sol                                     |        0.000970469 |            0           |            0 |
| array/push/array_push_nested_from_calldata.sol                            |       -0.000886423 |            0           |            0 |
| array/push/nested_bytes_push.sol                                          |        0.00111663  |            0           |            0 |
| array/push/array_push_struct_from_calldata.sol                            |        0.0218929   |           -0.00144935  |            0 |
| array/push/push_no_args_bytes.sol                                         |       -0.207675    |            0           |            0 |
| array/push/array_push_struct.sol                                          |        0.0155461   |            0           |            0 |
| inheritance/value_for_constructor.sol                                     |       -0.164684    |            0           |            0 |
| calldata/copy_from_calldata_removes_bytes_data.sol                        |       -0.000644616 |           -0.00193177  |            0 |
</details>

**Comment by ekpyron** [comment]
External test benchmarks don't look bad actually:

<details>
<summary>External test benchmarks</summary>

### `ir-no-optimize`
|          project | bytecode_size | deployment_gas | method_gas |
|:----------------:|--------------:|---------------:|-----------:|
|            brink |          `0%` |                |            |
|           colony |          `0%` |                |            |
|        elementfi |          `0%` |                |            |
|              ens |          `0%` |                |            |
|           gnosis |               |                |            |
|  perpetual-pools |          `0%` |          `-0%` |      `-0%` |
|          uniswap |          `0%` |          `-0%` |      `-0%` |
| yield_liquidator |          `0%` |          `+0%` |      `+0%` |
|         zeppelin |               |                |            |

### `ir-optimize-evm+yul`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|            brink | **`-0.08% ✅`** |                |                |
|           colony | **`+0.02% ❌`** |                |                |
|        elementfi | **`-0.21% ✅`** |                |                |
|              ens |           `0%` | **`+0.19% ❌`** |          `-0%` |
|           gnosis |                |                |                |
|  perpetual-pools | **`-0.19% ✅`** | **`-0.17% ✅`** | **`-0.02% ✅`** |
|          uniswap | **`-1.15% ✅`** | **`-0.88% ✅`** | **`-1.33% ✅`** |
| yield_liquidator | **`-0.19% ✅`** | **`-0.17% ✅`** |          `-0%` |
|         zeppelin |           `!V` |           `!V` |           `!V` |

### `ir-optimize-evm-only`
|          project | bytecode_size | deployment_gas | method_gas |
|:----------------:|--------------:|---------------:|-----------:|
|            brink |          `0%` |                |            |
|           colony |          `0%` |                |            |
|        elementfi |          `0%` |                |            |
|              ens |          `0%` |          `-0%` |       `0%` |
|           gnosis |               |                |            |
|  perpetual-pools |          `0%` |          `+0%` |      `+0%` |
|          uniswap |          `0%` |          `-0%` |      `-0%` |
| yield_liquidator |          `0%` |          `-0%` |       `0%` |
|         zeppelin |          `!V` |           `!V` |       `!V` |

### `legacy-no-optimize`
|          project | bytecode_size | deployment_gas |     method_gas |
|:----------------:|--------------:|---------------:|---------------:|
|            brink |          `0%` |                |                |
|           colony |          `0%` |                |                |
|        elementfi |          `0%` |                |                |
|              ens |          `0%` |                |                |
|           gnosis |          `0%` |          `-0%` |          `+0%` |
|  perpetual-pools |          `0%` |          `-0%` | **`+0.01% ❌`** |
|          uniswap |          `0%` |          `-0%` |          `+0%` |
| yield_liquidator |          `0%` |          `+0%` |           `0%` |
|         zeppelin |          `!V` |           `!V` |           `!V` |

### `legacy-optimize-evm+yul`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|            brink |           `0%` |                |                |
|           colony | **`+0.02% ❌`** |                |                |
|        elementfi | **`+0.05% ❌`** |                |                |
|              ens |           `0%` | **`+0.02% ❌`** |           `0%` |
|           gnosis | **`+0.02% ❌`** | **`+0.01% ❌`** | **`+0.21% ❌`** |
|  perpetual-pools | **`+0.01% ❌`** |          `+0%` | **`-0.01% ✅`** |
|          uniswap | **`+0.01% ❌`** | **`+0.01% ❌`** |          `+0%` |
| yield_liquidator | **`+0.03% ❌`** | **`+0.01% ❌`** |           `0%` |
|         zeppelin |           `!V` |           `!V` |           `!V` |

### `legacy-optimize-evm-only`
|          project | bytecode_size | deployment_gas |     method_gas |
|:----------------:|--------------:|---------------:|---------------:|
|            brink |          `0%` |                |                |
|           colony |          `0%` |                |                |
|        elementfi |          `0%` |                |                |
|              ens |          `0%` |          `-0%` |           `0%` |
|           gnosis |          `0%` |           `0%` |          `-0%` |
|  perpetual-pools |          `0%` |          `-0%` | **`+0.01% ❌`** |
|          uniswap |          `0%` |          `+0%` |          `-0%` |
| yield_liquidator |          `0%` |          `+0%` |          `-0%` |
|         zeppelin |          `!V` |           `!V` |           `!V` |


`!V` = version mismatch
`!B` = no value in the "before" version
`!A` = no value in the "after" version
`!T` = one or both values were not numeric and could not be compared
`-0` = very small negative value rounded to zero
`+0` = very small positive value rounded to zero

</details>

**Comment by nikola-matic** [comment]
> External test benchmarks don't look bad actually:
> External test benchmarks

Yeah, it really doesn't, at least for ir-optimized.

**Comment by ekpyron** [comment]
We should have a closer look into this in the context of https://github.com/ethereum/solidity/issues/14406 in any case - the results look a bit mixed in our semantics tests, but actually better in the external tests.
Interesting that the Yul snippets during legacy codegen seem to be adversely affected, but also not too badly.

So I'd be fine with merging this - then it'd need a Changelog entry and docs adjustments, though.

**Comment by nikola-matic** [comment]
> We should have a closer look into this in the context of #14406 in any case - the results look a bit mixed in our semantics tests, but actually better in the external tests. Interesting that the Yul snippets during legacy codegen seem to be adversely affected, but also not too badly.
> 
> So I'd be fine with merging this - then it'd need a Changelog entry and docs adjustments, though.

Dropping https://github.com/ethereum/solidity/issues/13817 here as well, since it's related (if we go for the `mu` approach).

**Comment by ekpyron** [comment]
So what about changelog and docs on this :-)?

**Comment by nikola-matic** [comment]
> So what about changelog and docs on this :-)?

Ah, you actually want this merged then (thought it was still a maybe). Will do.

File: Changelog.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ Language Features:
 
 Compiler Features:
  * Yul Optimizer: If ``PUSH0`` is supported, favor zero literals over storing zero values in variables.
+ * Yul Optimizer: Run ``Rematerializer`` and ``UnusedPruner`` steps as part of the default clean-up sequence.
```
Comment (ekpyron) [inline Changelog.md]:
```suggestion
 * Yul Optimizer: Run the ``Rematerializer`` and ``UnusedPruner`` steps at the end of the default clean-up sequence.
```

**Comment by ekpyron** [comment]
Should have been quicker in merging this - now the tests need a rebase

**Comment by nikola-matic** [comment]
> Should have been quicker in merging this - now the tests need a rebase

Completely forgot about this - even thought it was merged. I'll fix it today and ping you.

**Review by ekpyron — APPROVED** [review]

**Comment by ekpyron** [comment]
For the record: this is a bit of a tradeoff, this can have negative consequences and they do even show in the test expectations (e.g. [test/libyul/yulOptimizerTests/fullSuite/name_cleaner_reserved.yul](https://github.com/ethereum/solidity/pull/14468/files#diff-3b718776a58cf2052f0f8a3f254ef8374afaa62d5b2cd59f37cd4c6306404ea2) or [test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul](https://github.com/ethereum/solidity/pull/14468/files#diff-207df4c43759485ba550e494758de766ec7774a7f93fe1746a797923048ec27a) are not exactly good changes per se)

But it will decrease stack pressure and the real-world test data is positive, so let's give it a shot.

The cases I linked above we can actually keep in mind for https://github.com/ethereum/solidity/issues/13721, i.e. the code transform could take advantage whenever it knows that a value is already unconditionally on stack and it's reachable (the latter information we don't have on the Yul level unfortunately).

We can also look into this further in the context of https://github.com/ethereum/solidity/issues/14406

**Pull request merged**